File: cardDeck.py
##
#   Author: Akhil Bagaria
#   Date: November 16th, 2015
#   Coding challenge to manage a deck of cards. 
##
import copy
import sys

# Simple data structure representing a card. It has a unique identity
# and a pointer to another card
class Card:
    def __init__(self, identity, nextcard=None):
        self.identity = identity
        self.next = nextcard

    def __repr__(self):
        return str(self.identity)

# Simple data structure representing a deck of cards. It is a stack which can
# be pushed to both from the top and the bottom. It has a print function for 
# debugging purposes. 
class Deck:
    def __init__(self, numCards, topcard = None, bottomcard = None):
        self.numCards = numCards
        self.top = topcard
        self.bottom = bottomcard

    def __repr__(self):
        current = self.top
        if current is None:
            return ""
        returnstr = ""
        returnstr += str(current.identity) + '->'
        while current.next is not None:
            current = current.next
            returnstr += str(current.identity) + '->'
        return returnstr

# Call this function to initialize a deck of cards with numCards cards.
def makeDeck(numCards):
    # hand deck of n cards
    handdeck = Deck(numCards=numCards)

    # make deck in hand
    for i in range(numCards):
        card = Card(identity=i, nextcard=handdeck.top)

        # make the new card the top of the stack
        handdeck.top = card

        # make the first card the bottom of the stack
        if i == 0:
            handdeck.bottom = card

    return handdeck

# Given a deck of cards, simulate the process of picking out the 
# top card and placing that into the table deck. 
def moveToTop(handdeck, tabledeck):
    if handdeck.top == None:
        return 

    # if hand deck only has one card remaining, flag it so that we can 
    # later set its bottom to None
    bottomFlag = False
    if handdeck.top.identity == handdeck.bottom.identity:
        bottomFlag = True

    # top of the hand deck
    topcard = handdeck.top

    # update what is the top of the hand deck
    handdeck.top = handdeck.top.next

    # place the top of the hand deck to the top of the table deck
    topcard.next = tabledeck.top
    tabledeck.top = topcard

    # if topcard is the only card in the table deck, set its bottom
    if tabledeck.bottom is None:
        tabledeck.bottom = tabledeck.top

    # if top was also bottom (last card in deck), set the bottom to null
    if bottomFlag:
        handdeck.bottom = None

    return (handdeck, tabledeck)

# Move the top of the hand deck to the bottom
def moveToBottom(handdeck):

    # if no top card, return
    if handdeck.top is None:
        return

    # tmp vars for top and bottom cards
    # deep copy because python defaults to object bindings, not copies
    topcard =  copy.deepcopy(handdeck.top)
    bottomcard = copy.deepcopy(handdeck.bottom)

    # update the bottom of the deck
    handdeck.bottom.next = topcard
    handdeck.bottom = topcard
    handdeck.bottom.next = None

    # update the top of the deck
    handdeck.top = handdeck.top.next

    # return the updated deck
    return handdeck

def isEqual(originalHandDeck, tabledeck):
    if originalHandDeck.top is None:
        return None
    if tabledeck.top is None:
        return False

    tabledeck2         = copy.deepcopy(tabledeck)
    originalHandDeck2  = copy.deepcopy(originalHandDeck)

    currentTable = tabledeck2.top
    currentHand  = originalHandDeck2.top

    while currentHand.next is not None and currentTable.next is not None:
        if currentHand.identity != currentTable.identity:
            return False

        # update the currents
        currentHand  = currentHand.next
        currentTable = currentTable.next

    return True

# A round is defined as taking a card from hand and placing it on the table, 
# then placing the next card in hand to the bottom of the hand deck until all
# the cards in the hand deck is in on the table deck.
def playRound(handdeck, tabledeck):
    while handdeck.top is not None:
        moveToTop(handdeck, tabledeck)
        moveToBottom(handdeck)

if __name__ == '__main__':
    # get the number of cards as a terminal input
    numCards = int(sys.argv[1])

    # initialize the hand deck with numCards cards
    handdeck = makeDeck(numCards)

    # make a copy of the original deck for comparison later
    originalHandDeck = copy.deepcopy(handdeck)

    print "*** Starting Deck ***"
    print "hand deck = ", handdeck
    print "top of deck = ", handdeck.top
    print "bottom of deck = ", handdeck.bottom

    # make an empty table deck
    tabledeck = Deck(numCards=numCards)

    numRounds = 0
    while True:
        playRound(handdeck, tabledeck)
        print "table deck = ", tabledeck
        handdeck = copy.deepcopy(tabledeck)
        tabledeck = Deck(numCards)
        numRounds += 1
        if isEqual(originalHandDeck, handdeck):
            break
    print "Number of rounds to get back to original deck = ", numRounds 

File: card_deck.c
#include "card_deck.h"
#include <stdlib.h>

/*
*	This function takes in the number of cards and returns a pointer 
*	to the head and tail of the stack of cards.
*/
deck_t *make_deck(int num_cards) {

	// hand deck of n cards
	deck_t *deck = (deck_t *)malloc(sizeof(deck_t));

	deck->num_cards = num_cards;
	deck->top = NULL;
	deck->bottom = NULL;

	// make deck in hand
	int i;
	for (i = 0; i < num_cards; ++i) {
		card_t *card = (card_t *)malloc(sizeof(card_t));

		// assign unique identity to the card
		card->identity = i;

		// make the new card the top of the stack
		card->next = deck->top;

		// update the top of the deck
		deck->top = card;

		// make the bottom card the bottom of the stack
		if (i == 0) {
			deck->bottom = card;
			card->next = NULL;
		}
	}

	return deck;
}

/*
*	Given a hand deck and a table deck, move the top of the hand deck 
*	to the top of the table deck
*/
void move_to_top(deck_t *handdeck, deck_t *tabledeck) {
	
	// if hand deck is empty, there is nothing to move
	if (handdeck->top == NULL) {
		printf("move_to_top called on an empty deck\n");
		return;
	}

	// If hand deck only has one card remaning, then we need to make its 
	// bottom to nullptr. We thus need a flag to track if thats true
	int bottom_flag = 0;
	if (handdeck->top->identity == handdeck->bottom->identity) {
		bottom_flag = 1;
	}

	// store the 2nd card in a tmp var
	card_t *new_head = handdeck->top->next;

	// make the top of the hand deck point at the current top of the 
	// table deck as the its next card
	handdeck->top->next = tabledeck->top;

	// update the table deck's head
	tabledeck->top = handdeck->top;

	// update the hand deck's top
	handdeck->top = new_head;

	// if table deck only has one card, then set its bottom
	if (tabledeck->bottom == NULL) {
		tabledeck->bottom = tabledeck->top;
	}

	if (bottom_flag) {
		handdeck->bottom = NULL;
	}

	return;
}

/*
*	Given a pointer to the deck, move the top card to the bottom. 
* 	In the process, update the top and bottom of the deck.
*/
void move_to_bottom(deck_t *handdeck) {

	// if no card, then return
	if (handdeck->top == NULL) {
		return;
	}

	// if only one card in the deck, move_to_bottom is already done
	if (handdeck->top->next == NULL) {
		return;
	}

	// update bottom card's next from nullptr to top (the new bottom)
	handdeck->bottom->next = handdeck->top;

	// update the top and the bottom of the deck
	handdeck->bottom = handdeck->top;
	handdeck->top = handdeck->top->next;
	
	// update the new bottom card's next to be a nullptr instead of pointing to 
	// the 2nd card in the deck
	handdeck->bottom->next = NULL;

}

/*
*	Given two decks, return if they are the equal by looking at their 
*   respective card identities.
*/
int is_equal(deck_t *tabledeck) {
	
	if (tabledeck->top == NULL) {
		return -1;
	}

	card_t *current = tabledeck->top;
	int card_number = tabledeck->num_cards - 1;
	while (current->next != NULL) {
		if (current->identity != card_number) {
			return 0;
		}
		--card_number;
		current = current->next;
	}

	return 1;
}

void play_round(deck_t *handdeck, deck_t *tabledeck) {
	while (handdeck->top != NULL) {
		move_to_top(handdeck, tabledeck);
		move_to_bottom(handdeck);
	}
}

/*
*	Print function for a deck. Iterate through the cards 
*	in the deck, and print the identity of the cards to stdout
*/
void print(deck_t *deck) {
	
	card_t *current = deck->top;
	
	if (current == NULL)
		return;

	printf("%d ->", current->identity);

	while (current->next != NULL) {
		current = current->next;
		printf(" %d ->", current->identity);
	}

	printf("\n");

	return;
}

/*
*	Print function for a card so that top and bottom can easily be printed.
*/
void print_card(card_t *card) {
	if (card == NULL) {
		return;
	}
	printf("%d\n", card->identity);
	return;
}

int main(void) {

	int num_cards = 10;
	
	// make new deck and print it
	deck_t* handdeck = make_deck(num_cards);
	print(handdeck);

	// initialize empty table deck
	deck_t *tabledeck = (deck_t *)malloc(sizeof(deck_t));
	tabledeck->num_cards = num_cards;
	tabledeck->top = NULL;
	tabledeck->bottom = NULL;

	// play the game
	int num_rounds = 0;
	while (1) {
		play_round(handdeck, tabledeck);
		printf("table deck = ");
		print(tabledeck);
		// hand deck will get table deck as hand deck should be null after 
		// a round has been played
		handdeck->top = tabledeck->top;
		handdeck->bottom = tabledeck->bottom;

		// table deck is an empty deck again
		tabledeck->top = NULL;
		tabledeck->bottom = NULL;

		++num_rounds;
		if (is_equal(handdeck)) {
			break;
		}
	}
	printf("Number of rounds played = %d\n", num_rounds);


	return 1;
}